[enumeration] Add enumeration guide (#16)
TempleEight/temple-docs: TempleEight/docs/guide/enumeration.md

Temple's architecture makes it easy to define your own endpoints, as discussed in the [adding endpoints](adding-endpoints) guide, or even omit endpoints, again as discussed in the [omitting endpoints](omitting-endpoints) guide.

In addition to the basic functionality provided with the CRUD endpoints, we offer an additional `list` endpoint, which has different effects depending on other configurations provided in your Templefile.

We'll start with the most basic example, and then consider how the response changes, depending on access control and authentication settings.

## A basic list endpoint
In the [Getting Started](../getting-started) guide, we defined the following Templefile:

```
ExampleProject: project {
  #language(go);
  #database(postgres);
  #provider(dockerCompose);
}

ExampleService: service {
  foo: string;
  bar: int;
}
```

To add an additional `list` endpoint to our `ExampleService`, we can simply add the `#enumerable` metadata tag to that service:

```
ExampleProject: project {
  #language(go);
  #database(postgres);
  #provider(dockerCompose);
}

ExampleService: service {
  foo: string;
  bar: int;

  #enumerable;
}
```

> This generates a new endpoint, accessible at `/api/example-service/all`, and this will return all stored elements as a JSON list.

Reviewer: if we’re specifying it’s JSON, it’s called an array

Here, we'll assume you understand what it means to be using the `#authMethod` and `#auth` metadata in your project and services.

One advantage of using authentication is that we can restrict the response of the list endpoint to only include relevant information to the authentication token provided.
This means that we can restrict responses to only include entities that the authentication token has created.

For example, let's add authentication to our `ExampleProject`, by including a `User` service, and using `email` as the `#authMethod`:

```
ExampleProject: project {
  #language(go);
  #database(postgres);
  #provider(dockerCompose);
  #authMethod(email);
}

ExampleUser: service {
  name: string;

  #auth;
}

ExampleService: service {
  foo: string;
  bar: int;

  #enumerable;
}
```

Now when making requests to the `ExampleService` we'll need to include an authorization header.
Let's show an example of listing:

```bash
# Deploy the project locally
❯❯❯ source deploy.sh
...

# Create two access tokens, extract them from the JSON response, and assign each to an environment variable

> ❯❯❯ TOKEN1=$(curl -X POST $KONG_ENTRY/api/auth/register -d '{"email": "temple1@temple.com", "password":"abcdefgh"}' | jq -r .AccessToken)

Reviewer: Is installation of `jq` covered elsewhere? I don’t have it

Author: Ah good point, I'll mention it above this, as this is the first time it's used

# Create two new entities in the example service, one for each token
❯❯❯ curl -X POST $KONG_ENTRY/api/example-service -d '{"foo": "Hello", "bar": 123}' -H "Authorization: Bearer $TOKEN1"
{"id":"596d62db-8587-11ea-9d77-0242ac160003","foo":"Hello","bar":123}
❯❯❯ curl -X POST $KONG_ENTRY/api/example-service -d '{"foo": "World", "bar": 456}' -H "Authorization: Bearer $TOKEN2"
{"id":"60e013cd-8587-11ea-9d77-0242ac160003","foo":"World","bar":456}

# List all entities stored in the example service, using each token:
❯❯❯ curl -X GET $KONG_ENTRY/api/example-service/all -H "Authorization: Bearer $TOKEN1"
{
  "ExampleServiceList":[
    {"id":"596d62db-8587-11ea-9d77-0242ac160003","foo":"Hello","bar":123},
    {"id":"60e013cd-8587-11ea-9d77-0242ac160003","foo":"World","bar":456}
  ]
}

❯❯❯ curl -X GET $KONG_ENTRY/api/example-service/all -H "Authorization: Bearer $TOKEN2"
{
  "ExampleServiceList":[
    {"id":"596d62db-8587-11ea-9d77-0242ac160003","foo":"Hello","bar":123},
    {"id":"60e013cd-8587-11ea-9d77-0242ac160003","foo":"World","bar":456}
  ]
}
```

### Enumerating on `#auth` services

> :::important

Author: <img width="928" alt="Screenshot 2020-04-24 at 15 17 36" src="https://user-images.githubusercontent.com/16157582/80222439-bc17eb80-863e-11ea-8abe-7030bcb47a3f.png">

That's how you get this beauty - did it not render like that for you?

> :::important

Reviewer 2: Weird I couldn't get these to work when I tried in my stuff

> :::important

Author: You need to upgrade Docusaurus - have done that in the OpenAPI PR

Services which contain the `#auth` metadata are special, in that they provide a 1-to-1 mapping between an access token and an entity within that service.
This means that there can only be at most one entity for a given access token.

By default any service with a `#auth` block are implicitly `#readable(by: this)`, allowing only the access token that created the entity to read the information stored.
This means that it is not possible to have an `#enumerable` tag on an `#auth` service, since this would only ever return the list containing the entity for that access token.

> However, if the readability of the service is expanded to `#readable(by: all)`, usage of `#enumerable` is permitted, and will return all entities in that service.

Reviewer: “If you would like to add a list endpoint with your own logic for filtering the outputs, see our guide on adding endpoints.”?

> By default any service with a `#auth` block are implicitly `#readable(by: this)`, allowing only the access token that created the entity to read the information stored.
> This means that it is not possible to have an `#enumerable` tag on an `#auth` service, since this would only ever return the list containing the entity for that access token.
> 
> However, if the readability of the service is expanded to `#readable(by: all)`, usage of `#enumerable` is permitted, and will return all entities in that service.

Reviewer 3: ```suggestion
By default any service with a `#auth` block are implicitly `#readable(by: this)`, allowing only the access token that created the entity to read the information stored.
This means that without overriding to `readable(by: all)`, it is not possible to have an `#enumerable` tag on an `#auth` service, since this would only ever return the list containing the entity for that access token.
```
